File: 115.distinct-subsequences.cpp
class Solution {
public:
    int numDistinct(string s, string t) {
        if (s.empty() || t.empty() || s.length() < t.length()) {
            return 0;
        }
        vector<int> dp(t.length(), 0);
        if (s[0] == t[0]) {
            dp[0] = 1;
        }
        for (int i = 1; i < s.length(); i++) {
            for (int j = t.length() - 1; j > 0; j--) {
                if (s[i] == t[j]) {
                    dp[j] = dp[j] + dp[j-1];
                }
            }
            if (s[i] == t[0]) {
                dp[0]++;
            }
        }
        return dp[t.length() - 1];
    }
};

File: 115.distinct-subsequences.py
class Solution:
    def numDistinct(self, s, t):
        """
        :type s: str
        :type t: str
        :rtype: int
        """
        if not s or not t or len(s) < len(t):
            return 0
        dp = [0] * len(t)
        if s[0] == t[0]:
            dp[0] = 1
        for i in range(1, len(s)):
            for j in range(len(t) - 1, 0, -1):
                if s[i] == t[j]:
                    dp[j] = dp[j] + dp[j-1]
            if s[i] == t[0]:
                dp[0] += 1
        return dp[-1]
